src/main/java/org/breedinginsight/brapi/v2/dao/ProgramCache.java

```diff
@@ -0,0 +1,128 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.breedinginsight.brapi.v2.dao;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListenableFutureTask;
+import io.micronaut.http.server.exceptions.InternalServerException;
+import lombok.extern.slf4j.Slf4j;
+import org.brapi.client.v2.model.exceptions.ApiException;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.*;
+import java.util.function.Supplier;
+
+@Slf4j
+public class ProgramCache<R> {
+
+    private FetchFunction<UUID, R> fetchMethod;
+    Map<UUID, Semaphore> programSemaphore = new HashMap<>();
+
+    final Executor executor = Executors.newCachedThreadPool();
+    private LoadingCache<UUID, R> cache = CacheBuilder.newBuilder()
+            .build(new CacheLoader<>() {
+                @Override
+                public R load(UUID programId) throws Exception {
+                    try {
+                        return fetchMethod.apply(programId);
+                    } catch (Exception e) {
+                        log.error(e.getMessage());
+                        cache.invalidate(programId);
+                        throw e;
+                    }
+                }
+            });
+
+    public ProgramCache(FetchFunction fetchMethod, List<UUID> keys) {
+        this.fetchMethod = fetchMethod;
+        // Populate cache on start up
+        for (UUID key: keys) {
+            updateCache(key);
+        }
+    }
+
+    public ProgramCache(FetchFunction fetchMethod) {
+        this.fetchMethod = fetchMethod;
+    }
+
+    public R get(UUID programId) throws ApiException {
+        try {
+            // This will get current cache data, or wait for the refresh to finish if there is no cache data.
+            // TODO: Do we want to wait for a refresh method if it is running? Returns current data right now, even if old
+            if (!programSemaphore.containsKey(programId)) {
+                // For the case where the program cache isn't initialize yet. Could happen when a program is created
+                // and the germplasm is queried for the first time.
+                updateCache(programId);
+                return cache.get(programId);
+            } else if (cache.getIfPresent(programId) == null && programSemaphore.get(programId).availablePermits() == 1) {
+                // For the case where the cache is invalidated but hasn't been refreshed
+                updateCache(programId);
+                return cache.get(programId);
```

Comment:
It was just so there wouldn't be an extra unneeded cache refresh queued up. But, I removed it because it simplifies the code quite a bit removing it, so I think we can deal with an extra refresh in this edge case. Thanks for pointing this out!